Modules/_zstd/compressor.c

```diff
@@ -0,0 +1,670 @@
+/*
+Low level interface to Meta's zstd library for use in the `zstd` Python library.
+
+Original implementation by Ma Lin with patches by Rogdham.
+
+Refactored for the CPython standard library by Emma Harper Smith.
+*/
+
+/* ZstdCompressor class definitions */
+
+/*[clinic input]
+module _zstd
+class _zstd.ZstdCompressor "ZstdCompressor *" "clinic_state()->ZstdCompressor_type"
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=875bf614798f80cb]*/
+
+
+#ifndef Py_BUILD_CORE_BUILTIN
+#  define Py_BUILD_CORE_MODULE 1
+#endif
+
+#include "_zstdmodule.h"
+
+#include "buffer.h"
+
+#include <stddef.h>               // offsetof()
+
+/* -----------------------
+    ZstdCompressor code
+----------------------- */
+
+int
+_PyZstd_set_c_parameters(ZstdCompressor *self, PyObject *level_or_options,
+                         const char *arg_name, const char* arg_type)
+{
+    size_t zstd_ret;
+    _zstd_state* const _module_state = PyType_GetModuleState(Py_TYPE(self));
+    if (_module_state == NULL) {
+        return -1;
+    }
+
+    /* Integer compression level */
+    if (PyLong_Check(level_or_options)) {
+        const int level = PyLong_AsInt(level_or_options);
+        if (level == -1 && PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Compression level should be 32-bit signed int value.");
+            return -1;
+        }
+
+        /* Save for generating ZSTD_CDICT */
+        self->compression_level = level;
+
+        /* Set compressionLevel to compression context */
+        zstd_ret = ZSTD_CCtx_setParameter(self->cctx,
+                                          ZSTD_c_compressionLevel,
+                                          level);
+
+        /* Check error */
+        if (ZSTD_isError(zstd_ret)) {
+            set_zstd_error(_module_state, ERR_SET_C_LEVEL, zstd_ret);
+            return -1;
+        }
+        return 0;
+    }
+
+    /* Options dict */
+    if (PyDict_Check(level_or_options)) {
+        PyObject *key, *value;
+        Py_ssize_t pos = 0;
+
+        while (PyDict_Next(level_or_options, &pos, &key, &value)) {
+            /* Check key type */
+            if (Py_TYPE(key) == _module_state->DParameter_type) {
+                PyErr_SetString(PyExc_TypeError,
+                                "Key of compression option dict should "
+                                "NOT be DParameter.");
+                return -1;
+            }
+
+            /* Both key & value should be 32-bit signed int */
```

Comment:
Nit but if you want a 32-bit signed int, I suggest using `PyLong_AsInt32`, although I doubt this module will ever be part of an embedded Python on 16-bit archs.

Reply:
Hm, I *think* that the comment is a red herring and should be removed. The `ZSTD_CCtx_setParameter` function takes an `int`, not an `int32_t` and an enum value (so another `int`). Therefore I think the comment is wrong and should be updated.

Reply:
I removed this comment. In general I think it would be a better UX if we checked the bounds and report that rather than relying on the values  being `int32_t` or `int`. I left a TODO to add that.